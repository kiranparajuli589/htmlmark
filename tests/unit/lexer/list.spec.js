import Lexer from "../../../lib/lexer/index.js"


describe("list", () => {
	it("should tokenize a valid ordered list", () => {
		const lines = [
			"1. item **1**",
			"1. item [link](link-url)",
			"1. item 3 `code item`"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should tokenize a valid un-ordered list", () => {
		const lines = [
			"- item **1**",
			"- item [link](link-url)",
			"- item 3 `code item`"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should manage the list indent", () => {
		const lines = [
			"  - item **1**",
			"  - item [link](link-url)",
			"  - item 3 `code item`"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent[0].indent).toEqual(0)
	})
	it("less than 4 indent should not break the list", () => {
		const lines = [
			"  - item **1**",
			"  - item [link](link-url)",
			"- item 3 `code item`",
			"- item 4"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should tokenize the nested list", () => {
		const lines = [
			"- item **1**",
			"- item [link](link-url)",
			"    - item 3 `code item`",
			"        - item 4",
			"    - item 5 ~~strike~~"
		]

		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should tokenize list combination", () => {
		const lines = [
			"- one",
			"- two",
			"1. one",
			"2. two",
			"- [ ] c empty",
			"- [x] c checked",
			"1. [ ] c empty",
			"1. [x] c checked"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should allow lazy lexing", () => {
		const lines = [
			"- [x] list *item* one",
			"omg this is merged with the above line"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should allow lazy unless there is two line separator", () => {
		const lines = [
			"- one",
			"            two",
			"three",
			"",
			"	   ```",
			"	   here",
			"	   four",
			"	   ```",
			"",
			"- two",
			"- three"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("Should allow indented codeblock inside", () => {
		const lines = [
			"To put a code block within a list item, the code block needs",
			"to be indented *twice* -- 8 spaces or two tabs:",
			"*   A list item with a code block:",
			"",
			"        <code goes here>",
			"",
			"### Code Blocks"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it.each([
		{ l: [
			"- one",
			"-",
			"- two"
		] },
		{ l: [
			"-",
			"- one",
			"- two"
		] },
		{ l: [
			"- one",
			"- two",
			"-"
		] }
	])("should tokenize an empty list item", ({ l }) => {
		const lexer = new Lexer(l)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should tokenize a list with variable spaces beforehand", () => {
		const lines = `
  - item **1**
  - one
- two
- three
`
		const lexer = new Lexer(lines.split("\n"))
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should cope with tabbed indentations", () => {
		const lines = `3.  Is a blank line needed before an indented code block?
\t\t(\`Markdown.pl\` requires it, but this is not mentioned in the
\t\tdocumentation, and some implementations do not require it.)

\t\t\`\`\` markdown
\t\tparagraph
\t\t\t\tcode?
\t\t\`\`\`

5.  Can list markers be indented?  Can ordered list markers be right-aligned?

\t\t\`\`\` markdown
\t\t 8. item 1
\t\t 9. item 2
\t\t10. item 2a
\t\t\`\`\`

This is a paragraph.
`
		const lexer = new Lexer(lines.split("\n"))
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
})
