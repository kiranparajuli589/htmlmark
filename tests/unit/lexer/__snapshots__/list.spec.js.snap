// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list Should allow indented codeblock inside 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "To put a code block within a list item, the code block needs
to be indented *twice* -- 8 spaces or two tabs:",
    "tokens": Array [
      Object {
        "raw": "To put a code block within a list item, the code block needs to be indented ",
        "type": "text",
        "value": "To put a code block within a list item, the code block needs to be indented ",
      },
      Object {
        "raw": "*twice*",
        "tokens": Array [
          Object {
            "raw": "twice",
            "type": "text",
            "value": "twice",
          },
        ],
        "type": "italic",
      },
      Object {
        "raw": " -- 8 spaces or two tabs:",
        "type": "text",
        "value": " -- 8 spaces or two tabs:",
      },
    ],
    "type": "paragraph",
    "value": "To put a code block within a list item, the code block needs to be indented *twice* -- 8 spaces or two tabs:",
  },
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "*   A list item with a code block:

        <code goes here>",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "A list item with a code block:",
            "tokens": Array [
              Object {
                "raw": "A list item with a code block:",
                "type": "text",
                "value": "A list item with a code block:",
              },
            ],
            "type": "paragraph",
            "value": "A list item with a code block:",
          },
          Object {
            "type": "new-line",
          },
          Object {
            "indent": 8,
            "language": null,
            "raw": "        <code goes here>",
            "type": "code-block",
            "value": "<code goes here>",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "*",
      "ordered": false,
    },
    "raw": "*   A list item with a code block:

        <code goes here>",
    "type": "list",
  },
  Object {
    "type": "new-line",
  },
  Object {
    "indent": 0,
    "level": 3,
    "raw": "### Code Blocks",
    "setext": false,
    "tokens": Array [
      Object {
        "raw": "Code Blocks",
        "type": "text",
        "value": "Code Blocks",
      },
    ],
    "type": "heading",
    "value": "Code Blocks",
  },
]
`;

exports[`list less than 4 indent should not break the list 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "  - item **1**",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item **1**",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "**1**",
                "tokens": Array [
                  Object {
                    "raw": "1",
                    "type": "text",
                    "value": "1",
                  },
                ],
                "type": "bold",
              },
            ],
            "type": "paragraph",
            "value": "item **1**",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "  - item [link](link-url)",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item [link](link-url)",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "[link](link-url)",
                "tokens": Object {
                  "href": "link-url",
                  "title": Object {
                    "raw": "link",
                    "tokens": Array [
                      Object {
                        "raw": "link",
                        "type": "text",
                        "value": "link",
                      },
                    ],
                  },
                  "tooltip": undefined,
                },
                "type": "link",
              },
            ],
            "type": "paragraph",
            "value": "item [link](link-url)",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- item 3 \`code item\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item 3 \`code item\`",
            "tokens": Array [
              Object {
                "raw": "item 3 ",
                "type": "text",
                "value": "item 3 ",
              },
              Object {
                "raw": "\`code item\`",
                "type": "code",
                "value": "code item",
              },
            ],
            "type": "paragraph",
            "value": "item 3 \`code item\`",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- item 4",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item 4",
            "tokens": Array [
              Object {
                "raw": "item 4",
                "type": "text",
                "value": "item 4",
              },
            ],
            "type": "paragraph",
            "value": "item 4",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "  - item **1**
  - item [link](link-url)
- item 3 \`code item\`
- item 4",
    "type": "list",
  },
]
`;

exports[`list should allow lazy lexing 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": true,
        "count": null,
        "raw": "- [x] list *item* one
omg this is merged with the above line",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "list *item* one
omg this is merged with the above line",
            "tokens": Array [
              Object {
                "raw": "list ",
                "type": "text",
                "value": "list ",
              },
              Object {
                "raw": "*item*",
                "tokens": Array [
                  Object {
                    "raw": "item",
                    "type": "text",
                    "value": "item",
                  },
                ],
                "type": "italic",
              },
              Object {
                "raw": " one omg this is merged with the above line",
                "type": "text",
                "value": " one omg this is merged with the above line",
              },
            ],
            "type": "paragraph",
            "value": "list *item* one omg this is merged with the above line",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": true,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- [x] list *item* one
omg this is merged with the above line",
    "type": "list",
  },
]
`;

exports[`list should allow lazy unless there is two line separator 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- one
            two
three

	   \`\`\`
	   here
	   four
	   \`\`\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one
            two
three",
            "tokens": Array [
              Object {
                "raw": "one             two three",
                "type": "text",
                "value": "one             two three",
              },
            ],
            "type": "paragraph",
            "value": "one             two three",
          },
          Object {
            "type": "new-line",
          },
          Object {
            "indent": 4,
            "language": null,
            "raw": "	   \`\`\`
	   here
	   four
	   \`\`\`",
            "type": "code-block",
            "value": "here
four",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- three",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "three",
            "tokens": Array [
              Object {
                "raw": "three",
                "type": "text",
                "value": "three",
              },
            ],
            "type": "paragraph",
            "value": "three",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- one
            two
three

	   \`\`\`
	   here
	   four
	   \`\`\`
- two
- three",
    "type": "list",
  },
]
`;

exports[`list should cope with tabbed indentations 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": "3",
        "raw": "3.  Is a blank line needed before an indented code block?
		(\`Markdown.pl\` requires it, but this is not mentioned in the
		documentation, and some implementations do not require it.)

		\`\`\` markdown
		paragraph
				code?
		\`\`\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "Is a blank line needed before an indented code block?
		(\`Markdown.pl\` requires it, but this is not mentioned in the
		documentation, and some implementations do not require it.)",
            "tokens": Array [
              Object {
                "raw": "Is a blank line needed before an indented code block? 		(",
                "type": "text",
                "value": "Is a blank line needed before an indented code block? 		(",
              },
              Object {
                "raw": "\`Markdown.pl\`",
                "type": "code",
                "value": "Markdown.pl",
              },
              Object {
                "raw": " requires it, but this is not mentioned in the 		documentation, and some implementations do not require it.)",
                "type": "text",
                "value": " requires it, but this is not mentioned in the 		documentation, and some implementations do not require it.)",
              },
            ],
            "type": "paragraph",
            "value": "Is a blank line needed before an indented code block? 		(\`Markdown.pl\` requires it, but this is not mentioned in the 		documentation, and some implementations do not require it.)",
          },
          Object {
            "type": "new-line",
          },
          Object {
            "indent": 8,
            "language": "markdown",
            "raw": "		\`\`\` markdown
		paragraph
				code?
		\`\`\`",
            "type": "code-block",
            "value": "paragraph
		code?",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": "5",
        "raw": "5.  Can list markers be indented?  Can ordered list markers be right-aligned?

		\`\`\` markdown
		 8. item 1
		 9. item 2
		10. item 2a
		\`\`\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "Can list markers be indented?  Can ordered list markers be right-aligned?",
            "tokens": Array [
              Object {
                "raw": "Can list markers be indented?  Can ordered list markers be right-aligned?",
                "type": "text",
                "value": "Can list markers be indented?  Can ordered list markers be right-aligned?",
              },
            ],
            "type": "paragraph",
            "value": "Can list markers be indented?  Can ordered list markers be right-aligned?",
          },
          Object {
            "type": "new-line",
          },
          Object {
            "indent": 8,
            "language": "markdown",
            "raw": "		\`\`\` markdown
		 8. item 1
		 9. item 2
		10. item 2a
		\`\`\`",
            "type": "code-block",
            "value": " 8. item 1
 9. item 2
10. item 2a",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": null,
      "ordered": true,
    },
    "raw": "3.  Is a blank line needed before an indented code block?
		(\`Markdown.pl\` requires it, but this is not mentioned in the
		documentation, and some implementations do not require it.)

		\`\`\` markdown
		paragraph
				code?
		\`\`\`
5.  Can list markers be indented?  Can ordered list markers be right-aligned?

		\`\`\` markdown
		 8. item 1
		 9. item 2
		10. item 2a
		\`\`\`",
    "type": "list",
  },
  Object {
    "type": "new-line",
  },
  Object {
    "indent": 0,
    "raw": "This is a paragraph.",
    "tokens": Array [
      Object {
        "raw": "This is a paragraph.",
        "type": "text",
        "value": "This is a paragraph.",
      },
    ],
    "type": "paragraph",
    "value": "This is a paragraph.",
  },
  Object {
    "type": "new-line",
  },
]
`;

exports[`list should tokenize a list with variable spaces beforehand 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "  - item **1**",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item **1**",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "**1**",
                "tokens": Array [
                  Object {
                    "raw": "1",
                    "type": "text",
                    "value": "1",
                  },
                ],
                "type": "bold",
              },
            ],
            "type": "paragraph",
            "value": "item **1**",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "  - one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- three",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "three",
            "tokens": Array [
              Object {
                "raw": "three",
                "type": "text",
                "value": "three",
              },
            ],
            "type": "paragraph",
            "value": "three",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "  - item **1**
  - one
- two
- three",
    "type": "list",
  },
  Object {
    "type": "new-line",
  },
]
`;

exports[`list should tokenize a valid ordered list 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": "1",
        "raw": "1. item **1**",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item **1**",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "**1**",
                "tokens": Array [
                  Object {
                    "raw": "1",
                    "type": "text",
                    "value": "1",
                  },
                ],
                "type": "bold",
              },
            ],
            "type": "paragraph",
            "value": "item **1**",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": "1",
        "raw": "1. item [link](link-url)",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item [link](link-url)",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "[link](link-url)",
                "tokens": Object {
                  "href": "link-url",
                  "title": Object {
                    "raw": "link",
                    "tokens": Array [
                      Object {
                        "raw": "link",
                        "type": "text",
                        "value": "link",
                      },
                    ],
                  },
                  "tooltip": undefined,
                },
                "type": "link",
              },
            ],
            "type": "paragraph",
            "value": "item [link](link-url)",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": "1",
        "raw": "1. item 3 \`code item\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item 3 \`code item\`",
            "tokens": Array [
              Object {
                "raw": "item 3 ",
                "type": "text",
                "value": "item 3 ",
              },
              Object {
                "raw": "\`code item\`",
                "type": "code",
                "value": "code item",
              },
            ],
            "type": "paragraph",
            "value": "item 3 \`code item\`",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": null,
      "ordered": true,
    },
    "raw": "1. item **1**
1. item [link](link-url)
1. item 3 \`code item\`",
    "type": "list",
  },
]
`;

exports[`list should tokenize a valid un-ordered list 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- item **1**",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item **1**",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "**1**",
                "tokens": Array [
                  Object {
                    "raw": "1",
                    "type": "text",
                    "value": "1",
                  },
                ],
                "type": "bold",
              },
            ],
            "type": "paragraph",
            "value": "item **1**",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- item [link](link-url)",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item [link](link-url)",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "[link](link-url)",
                "tokens": Object {
                  "href": "link-url",
                  "title": Object {
                    "raw": "link",
                    "tokens": Array [
                      Object {
                        "raw": "link",
                        "type": "text",
                        "value": "link",
                      },
                    ],
                  },
                  "tooltip": undefined,
                },
                "type": "link",
              },
            ],
            "type": "paragraph",
            "value": "item [link](link-url)",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- item 3 \`code item\`",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item 3 \`code item\`",
            "tokens": Array [
              Object {
                "raw": "item 3 ",
                "type": "text",
                "value": "item 3 ",
              },
              Object {
                "raw": "\`code item\`",
                "type": "code",
                "value": "code item",
              },
            ],
            "type": "paragraph",
            "value": "item 3 \`code item\`",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- item **1**
- item [link](link-url)
- item 3 \`code item\`",
    "type": "list",
  },
]
`;

exports[`list should tokenize an empty list item 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "-",
        "tokens": Array [],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- one
-
- two",
    "type": "list",
  },
]
`;

exports[`list should tokenize an empty list item 2`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "-",
        "tokens": Array [],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "-
- one
- two",
    "type": "list",
  },
]
`;

exports[`list should tokenize an empty list item 3`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "-",
        "tokens": Array [],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- one
- two
-",
    "type": "list",
  },
]
`;

exports[`list should tokenize list combination 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- one
- two",
    "type": "list",
  },
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": "1",
        "raw": "1. one",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "one",
            "tokens": Array [
              Object {
                "raw": "one",
                "type": "text",
                "value": "one",
              },
            ],
            "type": "paragraph",
            "value": "one",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": "2",
        "raw": "2. two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "two",
            "tokens": Array [
              Object {
                "raw": "two",
                "type": "text",
                "value": "two",
              },
            ],
            "type": "paragraph",
            "value": "two",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": null,
      "ordered": true,
    },
    "raw": "1. one
2. two",
    "type": "list",
  },
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": false,
        "count": null,
        "raw": "- [ ] c empty",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "c empty",
            "tokens": Array [
              Object {
                "raw": "c empty",
                "type": "text",
                "value": "c empty",
              },
            ],
            "type": "paragraph",
            "value": "c empty",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": true,
        "count": null,
        "raw": "- [x] c checked",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "c checked",
            "tokens": Array [
              Object {
                "raw": "c checked",
                "type": "text",
                "value": "c checked",
              },
            ],
            "type": "paragraph",
            "value": "c checked",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": true,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- [ ] c empty
- [x] c checked",
    "type": "list",
  },
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": false,
        "count": "1",
        "raw": "1. [ ] c empty",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "c empty",
            "tokens": Array [
              Object {
                "raw": "c empty",
                "type": "text",
                "value": "c empty",
              },
            ],
            "type": "paragraph",
            "value": "c empty",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": true,
        "count": "1",
        "raw": "1. [x] c checked",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "c checked",
            "tokens": Array [
              Object {
                "raw": "c checked",
                "type": "text",
                "value": "c checked",
              },
            ],
            "type": "paragraph",
            "value": "c checked",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": true,
      "identifier": null,
      "ordered": true,
    },
    "raw": "1. [ ] c empty
1. [x] c checked",
    "type": "list",
  },
]
`;

exports[`list should tokenize the nested list 1`] = `
Array [
  Object {
    "indent": 0,
    "items": Array [
      Object {
        "checked": null,
        "count": null,
        "raw": "- item **1**",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item **1**",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "**1**",
                "tokens": Array [
                  Object {
                    "raw": "1",
                    "type": "text",
                    "value": "1",
                  },
                ],
                "type": "bold",
              },
            ],
            "type": "paragraph",
            "value": "item **1**",
          },
        ],
        "type": "list-item",
      },
      Object {
        "checked": null,
        "count": null,
        "raw": "- item [link](link-url)
    - item 3 \`code item\`
        - item 4
    - item 5 ~~strike~~",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": "item [link](link-url)",
            "tokens": Array [
              Object {
                "raw": "item ",
                "type": "text",
                "value": "item ",
              },
              Object {
                "raw": "[link](link-url)",
                "tokens": Object {
                  "href": "link-url",
                  "title": Object {
                    "raw": "link",
                    "tokens": Array [
                      Object {
                        "raw": "link",
                        "type": "text",
                        "value": "link",
                      },
                    ],
                  },
                  "tooltip": undefined,
                },
                "type": "link",
              },
            ],
            "type": "paragraph",
            "value": "item [link](link-url)",
          },
          Object {
            "indent": 4,
            "items": Array [
              Object {
                "checked": null,
                "count": null,
                "raw": "    - item 3 \`code item\`
        - item 4",
                "tokens": Array [
                  Object {
                    "indent": 0,
                    "raw": "item 3 \`code item\`",
                    "tokens": Array [
                      Object {
                        "raw": "item 3 ",
                        "type": "text",
                        "value": "item 3 ",
                      },
                      Object {
                        "raw": "\`code item\`",
                        "type": "code",
                        "value": "code item",
                      },
                    ],
                    "type": "paragraph",
                    "value": "item 3 \`code item\`",
                  },
                  Object {
                    "indent": 8,
                    "items": Array [
                      Object {
                        "checked": null,
                        "count": null,
                        "raw": "        - item 4",
                        "tokens": Array [
                          Object {
                            "indent": 0,
                            "raw": "item 4",
                            "tokens": Array [
                              Object {
                                "raw": "item 4",
                                "type": "text",
                                "value": "item 4",
                              },
                            ],
                            "type": "paragraph",
                            "value": "item 4",
                          },
                        ],
                        "type": "list-item",
                      },
                    ],
                    "meta": Object {
                      "checklist": false,
                      "identifier": "-",
                      "ordered": false,
                    },
                    "raw": "        - item 4",
                    "type": "list",
                  },
                ],
                "type": "list-item",
              },
              Object {
                "checked": null,
                "count": null,
                "raw": "    - item 5 ~~strike~~",
                "tokens": Array [
                  Object {
                    "indent": 0,
                    "raw": "item 5 ~~strike~~",
                    "tokens": Array [
                      Object {
                        "raw": "item 5 ",
                        "type": "text",
                        "value": "item 5 ",
                      },
                      Object {
                        "raw": "~~strike~~",
                        "tokens": Array [
                          Object {
                            "raw": "strike",
                            "type": "text",
                            "value": "strike",
                          },
                        ],
                        "type": "strike-through",
                      },
                    ],
                    "type": "paragraph",
                    "value": "item 5 ~~strike~~",
                  },
                ],
                "type": "list-item",
              },
            ],
            "meta": Object {
              "checklist": false,
              "identifier": "-",
              "ordered": false,
            },
            "raw": "    - item 3 \`code item\`
        - item 4
    - item 5 ~~strike~~",
            "type": "list",
          },
        ],
        "type": "list-item",
      },
    ],
    "meta": Object {
      "checklist": false,
      "identifier": "-",
      "ordered": false,
    },
    "raw": "- item **1**
- item [link](link-url)
    - item 3 \`code item\`
        - item 4
    - item 5 ~~strike~~",
    "type": "list",
  },
]
`;
